Dynamic Commands

Dosx001 opened the pull request:
This change will allow aliasme.sh to run dynamic commands instead of just static commands. For example, when you create something like this "al add new 'cd'". Then when you run "al new root/directory" then it will change your directory to "../root/directory" or anything base on your input.

Dosx001 committed:
This chagne will allow aliasme.sh to run dynamic commands instead of
just static commands. For example when you create something like this "al add
new 'cd'". Then when you run "al new root/directory" then it will change your
directory to "../root/directory" or anything base on your input.

Jintin commented:
It's really cool feature that I also want to include for aliasme for a long time.
Just some though to add on, can we indicate the dynamic part to place at anywhere we want of the command?
Also is it possible to make support multiple dynamic part?
Thanks for you contribute!!

Dosx001 commented:
What do you mean anywhere? Do you mean something like this "al root/directory new"?

I think for multiple dynamic parts, to keep the code simple, we could have a keyword to start a multiple dynamic command. Something like this "al multi_dynamic alias  input1 input2".

Dosx001 committed:
This change will allow for multiple parts dynamic commands

Dosx001 commented:
MD stands for multiple dynamic
"al dynamic" is just "vim" 
![image](https://user-images.githubusercontent.com/62494922/103484226-eb702200-4da1-11eb-9a64-329e041d6e57.png)

Dosx001 committed:
Clean up code

Dosx001 committed:
Instead adding onto a alias this change will allow you to also create alias where
you can change its' contents per call

Dosx001 commented:
Ya am getting a little crazy with this. When you call WordCount the "?" is replaced by whatever put as an argument on the command line.
![image](https://user-images.githubusercontent.com/62494922/103488826-d789e780-4dc4-11eb-9f3d-6bc3faf2c930.png)
With a little work, I can make a combination of both the Multiple Dynamic and Fill

Dosx001 committed:
Added multi fill option

Dosx001 commented:
This one is pretty cool. Oh ya, I changed "Fill" to just "F" and "MF" stands for Multiple Fill.
![image](https://user-images.githubusercontent.com/62494922/103490616-f17df700-4dd1-11eb-83b0-fdcaeb4ee4eb.png)

Dosx001 committed:
Removed the need to distinguish between a dynamic and multi dynamic alias

Dosx001 committed:
Every alias now has cmdType Default, Dynamic, Fill, or Multi Fill. Some
functions were modify to accommodate for cmdType. However autocomplete is broken.

Dosx001 committed:
Remove() is fixed

Dosx001 committed:
Fixed autocomplete on the command line

Jintin commented:
> What do you mean anywhere? Do you mean something like this "al root/directory new"?
> 
> I think for multiple dynamic parts, to keep the code simple, we could have a keyword to start a multiple dynamic command. Something like this "al multi_dynamic alias input1 input2".

Ya, similar to what you say, put a symbol to indicate the dynamic part.
Don't know why I don't get notify about the following chat, you can pin my id next time.

Dosx001 committed:
Fixed add()

Dosx001 committed:
All Fills and Multi Fills are now just Fills. While the new Multi Fill is now is a combination of Fill and Dynamic

Dosx001 commented:
@Jintin 
"""Some suggest on the dynamic or multi-fill part.
Is is possible to put some indicator like $1 $2 indicate multiple usage? or $0 for list type of data?
example:
mv $1.md $1.txt
vim $0"""

I did use $1, $2, etc before. But you would need to backlash it when adding a new alias. Something like this "al add new '\\$1 \\$2 \\$3' FM". I didn't like having to type an extra character so I changed "$" to "?".

Dosx001 commented:
@Jintin Check out these new changes!!! This pull request just keeping getting crazier lol.

First off, the biggest change is now Fills and Multi Fills are now the same thing just Fills. While the new Multi Fills is now a combination of Fill and Dynamic. 

Here is a visual example.
![image](https://user-images.githubusercontent.com/62494922/103582467-c27b8a00-4e92-11eb-81e7-aff75e2aa602.png)

When you “al ls”, you’ll see that Multi Fills have a number next to them, this number is the amount of arguments needed to run the alias command. In order to add a Multi Fill alias you need to do something like this “al add VP ‘vim ?1 ?2 ?3 ?4 -p’ MF 4”. Then just run like this “al VP a.py b.py c.py d.py”

You can also have something like this.
![image](https://user-images.githubusercontent.com/62494922/103592257-6e2ed500-4ea7-11eb-830f-7847101535a9.png)

Dosx001 committed:
Fills can now take multiple arguments and changed the declaration for each cmdType in add()

Dosx001 commented:
Fills can now take as many arguments as possible like this

![image](https://user-images.githubusercontent.com/62494922/103593845-c9fb5d00-4eab-11eb-89e5-1a794498f4df.png)

I also changed how you declare the cmdType when adding a new alias. I kept typing "FM" instead of "MF", so this change will greatly help me lol, and overall less typing. 
D -> d
F -> f
MF -> m

Dosx001 committed:
Code clean up

Jintin commented:
Thanks for contribution, I know it been a while but I'm back on this project. Currently we've added the brew support(feel free to try) and more test. And it make the current pr conflict to continue. I'll close it for now since it's been really a long time but feel free to raise another one if you're still interested in this. And here are some other suggestion I got from AI assistent which might be helpful:

  1. Breaking Change: Storage Format
  The change from a 2-line format (name, command) to a 3-line format (name,
  command, type) in ~/.aliasme/cmd is a breaking change.
   - Issue: Existing aliases for all current users will break. _list, _remove,
     _excute, and _bashauto will misread the file because they expect three
     lines per entry but existing files only have two.
   - Recommendation: Implement a migration script or update the parsing logic to
     be backward compatible (e.g., if the third line is missing, default to
     "Default" type).

  2. Hardcoded Paths
  The PR uses hardcoded paths like ~/.aliasme/cmd.
   - Issue: The current version of aliasme.sh uses variables like ALIASME_CMD
     and ALIASME_DIR to allow customization and cleaner code.
   - Recommendation: Use the existing variables instead of hardcoding the path.

  3. Argument Quoting and eval
  The way arguments are handled for "Dynamic", "Fill", and "Multi Fill" is prone
  to issues with spaces.
   - Issue: 
      ```
      cmds+=$arg\ 
      eval $line $cmds
      ```
    If an argument contains a space, it will be split when passed to eval.
   - Recommendation: Use Bash arrays to handle arguments safely. For "Dynamic"
     mode, you can simply use "$@" (after shifting the command name) to pass all
     remaining arguments preserved with their quoting.

  4. ZSH Support
  The PR updates the Bash completion logic (_bashauto) but neglects the ZSH
  logic.
   - Issue: ZSH users will experience broken completion because _autocomplete
     still expects the 2-line format.
   - Recommendation: Update the ZSH branch in _autocomplete to also skip the
     third line when building the opts list.

  5. Code Quality and Consistency
   - Indentation: The PR introduces inconsistent indentation (mix of tabs and
     spaces). It’s best to stick to the project's existing style (looks like
     tabs or 4 spaces).
   - Bash-isms: The use of seq is fine but for ((i=2; i<=$#; i++)) is often
     preferred in pure Bash. The array assignment cmds=() followed by
     string-like concatenation cmds+=$arg\  is confusing and likely incorrect
     for true array usage.

  6. Missing Test Updates
  The existing tests in test/aliastest.sh will fail because they check for the
  old _list output format.
   - Recommendation: Update the tests to reflect the new output format and add
     new test cases for the new command types.

  7. "Multi Fill" Logic
  The logic for splitting cmdType to get the number of fills is a bit brittle:
  ```
    fills=($cmdType)
    for i in $(seq 1 ${fills[2]}); do
   ```
  If cmdType is "Multi Fill 3", fills[2] is "3". This works but assumes the
  string format perfectly. A more robust way would be to store the count
  separately or parse it more explicitly.
